KAFKA-16324: Move BrokerApiVersionsCommand to tools

FrankYang0529 opened the pull request:
* Migrate `BrokerApiVersionsCommand` to tools module.
* Migrate `BrokerApiVersionsCommandTest` to user `ClusterTestExtensions`.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

chia7712 reviewed (commented):
@FrankYang0529 could you please update the scripts too?

https://github.com/apache/kafka/blob/trunk/bin/kafka-broker-api-versions.sh#L17
https://github.com/apache/kafka/blob/trunk/bin/windows/kafka-broker-api-versions.bat#L17

chia7712 commented on tools/src/main/java/org/apache/kafka/tools/BrokerApiVersionsCommand.java:
```diff
@@ -0,0 +1,337 @@
+                nodes = findAllBrokers();
+                if (nodes.isEmpty()) {
+                    try {
+                        Thread.sleep(50);
```
`TimeUnit.MILLISECONDS.sleep(50);`

chia7712 commented on tools/src/main/java/org/apache/kafka/tools/BrokerApiVersionsCommand.java:
```diff
@@ -0,0 +1,337 @@
+                if (nodes.isEmpty()) {
+                    try {
+                        Thread.sleep(50);
+                    } catch (InterruptedException e) {
```
not sure why we need to catch it here. could you please throw it directly as it is a command-line tool

chia7712 commented on tools/src/main/java/org/apache/kafka/tools/BrokerApiVersionsCommand.java:
```diff
@@ -0,0 +1,337 @@
+        private final List<Node> bootstrapBrokers;
+        private final ConcurrentLinkedQueue<RequestFuture<ClientResponse>> pendingFutures = new ConcurrentLinkedQueue<>();
+
+        static class AdminConfig extends AbstractConfig {
```
Could you please remove it as we can use `AbstractConfig` directly? for instance:
```java
        static AdminClient create(Map<String, ?> props) {
            return create(new AbstractConfig(ADMIN_CONFIG_DEF, props, false));
        }
```

chia7712 commented on tools/src/main/java/org/apache/kafka/tools/BrokerApiVersionsCommand.java:
```diff
@@ -0,0 +1,337 @@
+            return create(Utils.propsToMap(props));
+        }
+
+        static AdminClient create(Map<String, ?> props) {
```
Do we need those constructors?

chia7712 commented on tools/src/main/java/org/apache/kafka/tools/BrokerApiVersionsCommand.java:
```diff
@@ -0,0 +1,337 @@
+
+        private AbstractResponse send(Node target, AbstractRequest.Builder<?> request) throws InterruptedException {
+            RequestFuture<ClientResponse> future = client.send(target, request);
+            pendingFutures.add(future);
```
Do we need another thread to poll client? Maybe we can poll the client by the same thread to simplify this tool?

chia7712 reviewed (commented):
@FrankYang0529 thanks for this nice migration

FrankYang0529 commented:
Hi @chia7712, thanks for review. I addressed all comments. Failed cases in CI can't be reproduced on my laptop. Could you take a look when you have time? Thank you.

chia7712 commented on tools/src/main/java/org/apache/kafka/tools/BrokerApiVersionsCommand.java:
```diff
@@ -0,0 +1,296 @@
+
+        private AbstractResponse send(Node target, AbstractRequest.Builder<?> request) {
+            RequestFuture<ClientResponse> future = client.send(target, request);
+            while (!future.isDone()) {
```
Should we handle the timeout manually?

FrankYang0529 replied:
I add a test case `testAdminSendNoBlock` to confirm `client.send` doesn't block forever. Thank you.

chia7712 commented on tools/src/main/java/org/apache/kafka/tools/BrokerApiVersionsCommand.java:
```diff
@@ -0,0 +1,296 @@
+        private KafkaFuture<NodeApiVersions> getNodeApiVersions(Node node) {
+            final KafkaFutureImpl<NodeApiVersions> future = new KafkaFutureImpl<>();
+            try {
+                ApiVersionsResponse response = (ApiVersionsResponse) send(node, new ApiVersionsRequest.Builder());
```
That will send the API_VERSION request to brokers one by one. Could you please leverage the `Future` (or `CompletableFuture`) to speedup it?

FrankYang0529 replied:
Hi @chia7712, original class sends request one by one. I think we can keep the behavior and do some refactor in the future if someone complains about this. Thanks.

https://github.com/apache/kafka/blob/c5e91546721b6b160732667ca7450d22927cc3b3/core/src/main/scala/kafka/admin/BrokerApiVersionsCommand.scala#L136-L138

chia7712 reviewed (commented):
@FrankYang0529 thanks for updates. a couple of comments are left.

chia7712 reviewed (commented):
@FrankYang0529 Could you please update the class path of following scripts?

https://github.com/apache/kafka/blob/trunk/bin/kafka-broker-api-versions.sh
https://github.com/apache/kafka/blob/trunk/bin/windows/kafka-broker-api-versions.bat

FrankYang0529 commented:
> @FrankYang0529 Could you please update the class path of following scripts?
> 
> https://github.com/apache/kafka/blob/trunk/bin/kafka-broker-api-versions.sh https://github.com/apache/kafka/blob/trunk/bin/windows/kafka-broker-api-versions.bat

Thank you. Updated it.

chia7712 commented on tools/src/test/java/org/apache/kafka/tools/BrokerApiVersionsCommandTest.java:
```diff
@@ -0,0 +1,107 @@
+ */
+package org.apache.kafka.tools;
+
+//import kafka.server.KafkaBroker;
```
could you please remove this?

FrankYang0529 replied:
Sorry for that. Updated it.

chia7712 reviewed (commented):
@FrankYang0529 thanks for updates. one tiny comment is left

FrankYang0529 committed:
KAFKA-16324: Move BrokerApiVersionsCommand to tools

Signed-off-by: PoAn Yang <payang@apache.org>

chia7712 reviewed (approved):
LGTM
